fdbserver/masterserver.actor.cpp

```diff
@@ -147,11 +154,46 @@ ACTOR Future<Void> getVersion(Reference<MasterData> self, GetCommitVersionReques
 			if (BUGGIFY) {
 				t1 = self->lastVersionTime;
 			}
-			rep.prevVersion = self->version;
-			self->version +=
+
+			// Versions should roughly follow wall-clock time, based on the
+			// system clock of the current machine and an FDB-specific epoch.
+			// Calculate the expected version and determine whether we need to
+			// hand out versions faster or slower to stay in sync with the
+			// clock.
+			Version toAdd =
 			    std::max<Version>(1,
 			                      std::min<Version>(SERVER_KNOBS->MAX_READ_TRANSACTION_LIFE_VERSIONS,
 			                                        SERVER_KNOBS->VERSIONS_PER_SECOND * (t1 - self->lastVersionTime)));
+			Version expectedVersion = g_network->timer() * SERVER_KNOBS->VERSIONS_PER_SECOND - self->referenceVersion;
+
+			// How far the new version is off from the expected version.
+			Version offset = self->version + toAdd - expectedVersion;
+			// Clamp the offset to a specific range. This prevents huge version
+			// jumps.
+			Version clamped =
+			    std::clamp(offset, -SERVER_KNOBS->MAX_VERSION_RATE_OFFSET, SERVER_KNOBS->MAX_VERSION_RATE_OFFSET);
+
+			// Translate the version offset (in the range [-1,000,000,
+			// 1,000,000]) to a value in the range [0.9, 1.1]). Note that these
+			// values are examples, and are based on the
+			// MAX_VERSION_RATE_OFFSET and MAX_VERSION_RATE_MODIFIER knobs.
+			double modifier =
+			    ((1.0 + SERVER_KNOBS->MAX_VERSION_RATE_MODIFIER) - (1.0 - SERVER_KNOBS->MAX_VERSION_RATE_MODIFIER)) /
+			    (2 * SERVER_KNOBS->MAX_VERSION_RATE_OFFSET);
+			// Use std::pow to apply a more aggressive curve to version
+			// corrections. Versions further from the expected version will
+			// receive a larger multiplier.
+			double versionCorrection =
+			    std::clamp(std::pow((1.0 - SERVER_KNOBS->MAX_VERSION_RATE_MODIFIER) +
+			                            modifier * (-clamped + SERVER_KNOBS->MAX_VERSION_RATE_OFFSET),
+			                        2.5),
+			               1.0 - SERVER_KNOBS->MAX_VERSION_RATE_MODIFIER,
+			               1.0 + SERVER_KNOBS->MAX_VERSION_RATE_MODIFIER);
+			;
```

Comment:
I think your version of just jumping exactly to the expected version if the jump is small enough makes more sense. The only potential issue could be if the system clock gets "jumpy" or something and jumps back and forth in time. But the only real problem this would cause is that the version may increment by at most 1 version each time.

Reply:
I thought about this a little more, and I think the main motivation behind the original version is to keep the rate of versions given out relatively consistent, around a million per second. But now, it can be varied by a multiplier up to 90% or 110% of the current rate. Maybe this doesn't matter so much if versions now follow wall clock time, but at least to me I think it makes sense to continue giving out versions as consistently as possible, and simply speed up or slow down to try and match what the expected version is.

Reply:
I see, this probably does have the effect of smoothing out the changes. If there is some variance in the offset that centers around 0, then you would be making smaller corrections but achieving an otherwise similar result. If there is a big change, then you are going to jump to 90% or 110% straight away as well.

One case that I'm less certain about is what happens if time starts advancing at 99.5% speed or something. In the simple implementation, versions would advance at the same rate as time (as long as it's within the 10% window). In the smoothed implementation, you would advance closer to 100% initially until the time drift increases, and probably then overshoot the actual rate. Then you would either oscillate or converge on the right rate, though possibly the convergence happens on an offset that isn't 0. That is to say, version and time might be advancing at the same 99.5% speed but with a non-zero gap between them.